### ISSUE TITLE
Retrieving host information on all Linode instances

### ISSUE THREAD

**Issue by lwm** [linode/linode_api4-python#141]
Again, building the Ansible dynamic inventory and need to collect 'hostvars' which are values against hosts which help to identify them or give further information about the instance. I'm so far doing something like:

```python
        for instance in self.instances:
            hostvars = [
                {'name': 'id', 'value': instance.id},
                {'name': 'ipv4', 'value': instance.ipv4},
                {'name': 'ipv6', 'value': instance.ipv6},
                {'name': 'region', 'value': instance.region.id},
                {'name': 'type', 'value': instance.type.id},
            ]

            for hostvar in hostvars:
                self.inventory.set_variable(
                    instance.label,
                    hostvar['name'],
                    hostvar['value'],
                )
```

This is OK for now but 1) is missing data (see below about properties) 2) won't include any new information as the instance API evolves and will continue to need updates. 

I see that `instance.properties` exists (and looks to have **everything** I might need but I am not sure how to use that as I don't really undertand the purpose of `Property` class). If I could get that payload, serialized, that would be great. I currently see:

```
ipdb> i1.properties
{'status': <linode_api4.objects.base.Property instance at 0x7ff938ca2b48>, 'updated': <linode_api4.objects.base.Property instance at 0x7ff938ca2bd8>, 'disks': <linode_api4.objects.base.Property instance at 0x7ff938ca2cf8>, 'group': <linode_api4.objects.base.Property instance at 0x7ff938ca2b00>, 'created': <linode_api4.objects.base.Property instance at 0x7ff938ca2b90>, 'hypervisor': <linode_api4.objects.base.Property instance at 0x7ff938ca2e18>, 'region': <linode_api4.objects.base.Property instance at 0x7ff938ca2c20>, 'alerts': <linode_api4.objects.base.Property instance at 0x7ff938ca2c68>, 'label': <linode_api4.objects.base.Property instance at 0x7ff938ca2ab8>, 'specs': <linode_api4.objects.base.Property instance at 0x7ff938ca2e60>, 'backups': <linode_api4.objects.base.Property instance at 0x7ff938ca2dd0>, 'ipv4': <linode_api4.objects.base.Property instance at 0x7ff938c85680>, 'image': <linode_api4.objects.base.Property instance at 0x7ff938ca2cb0>, 'ipv6': <linode_api4.objects.base.Property instance at 0x7ff938c85638>, 'configs': <linode_api4.objects.base.Property instance at 0x7ff938ca2d40>, 'type': <linode_api4.objects.base.Property instance at 0x7ff938ca2d88>, 'id': <linode_api4.objects.base.Property instance at 0x7ff938ca2a70>}
```

**Comment by lwm** [comment]
Another issue that this touches on, when writing the new Linode API v4 module is that I need a serialised format for the `Instance` class which I currently (AFAIK) have to piece together manually and access all the instance properties that I think might be useful. A `dict(instance)` -> `{'label: 'foo', ....}` tranformer would be great! I'll poke around and see if that is possible now.

**Comment by lwm** [comment]
Right, I've seen now `instance._serialize` but it isn't really showing me all the info I need.

**Comment by Dorthu** [comment]
Hmm..that's a tough one.  The `_serialize` method is a private helper for saving updated, editable values (so it only returns mutable properties).  The `properties` dict is a class-level construct that gives the library hints about what the API should return for this object.  It could probably be used to re-JSONify the object, but nothing exists to do so currently.  It gets a little harder because I'm doing some transformation on some properties (for example `Instance.region` is returned by the API as an id, like "us-east", but the library turns it into a `Region` object representing that ID).  Related collections (`Instance.disks` for example) get even more complicated.

Something like this might make it easier, but it's still not going to properly handle related elements:

```python
In [18]: def make_dict(inst, keys):
    ...:     ret = {}
    ...:     for k in keys:
    ...:         ret[k] = getattr(inst, k)
    ...:     return ret
    ...:

In [19]: make_dict(inst, ('id','label','group','ipv4','ipv6','region','image'))
Out[19]:
{'group': '',
 'id': 10337453,
 'image': Image: linode/ubuntu17.10,
 'ipv4': ['69.164.211.83'],
 'ipv6': '2600:3c03::f03c:91ff:fe22:88dc/64',
 'label': 'linode10337453',
 'region': Region: us-east}
```

How important is this to ansible?  If you need a generic way to get the JSON that created objects (without all the frills the library adds), I could work something in for it.

**Comment by lwm** [comment]
I see, thanks for this explanation and examples! 

> How important is this to ansible? If you need a generic way to get the JSON that created objects (without all the frills the library adds), I could work something in for it.

Pretty much the bread and butter the module will run off ;) 

They are used for filtering instances when building an inventory of machines for running tasks. You might want to run a certain command on instances that have image = foo and region = bar. More importantly though, when you create/delete/whatever with an instance from the module we want to hand back a JSON serialised payload of all the relevant information of the instance that can be acted on by another task.

Finally, (and this probably comes later), there will hopefully be a `linode_v4_facts` which just hands back a load of JSON serialised information about the user's Linode account, instances, DNS, SSH keys, etc. So, I suppose I'll be nagging about this more and more ;)

Would be great to see if you come with something general! Only needed for Linode instances now.

I'll use your property accessing function for now :smile:

**Comment by Dorthu** [comment]
@lwm does the solution in #143 do what you need it to do?

**Comment by lwm** [comment]
Checking now! Thanks a lot.

**Comment by Dorthu** [comment]
These changes (and a bunch of others) have shipped in version `2.0.0` (the breaking change was related to the OAuth exchange and shouldn't impact your work).  I'm going to close this since the `_raw_json` field is now available in a mainline release.
